blockchain-albatross/src/blockchain.rs

```diff
@@ -592,6 +599,44 @@ impl Blockchain {
                 return Err(PushError::InvalidBlock(BlockError::InvalidJustification));
             }
 
+            // Check if there are two blocks in the same slot. Since we already
+            // verified the validator for the current slot, this is enough to
+            // check for fork proofs. A fork proof doesn't apply to the genesis
+            // block, so skip block 0.
+            if block.block_number() != 0 {
+                // Count the microblocks after the last macroblock
+                let microblocks: Vec<Block> = self.chain_store.get_blocks_at(block.block_number(), false, Some(&read_txn));
+
+                // Get the microheader from the block
+                let microheader1 = block.clone().unwrap_micro().header;
+
+                // Get the justification for the block. We asume that the
+                // validator's signature is valid.
+                let justification1 = block.clone().unwrap_micro().justification.signature;
+
+                // Get the view number from the block
+                let view_number = block.clone().view_number();
+
+                for microblock in microblocks.iter() {
+                    let (microblock_slot, _) = self.get_slot_at(microblock.block_number(), microblock.view_number(), Some(&read_txn)).unwrap();
+
+                    // If there's another microblock set to this slot, we
+                    // notify the fork event.
+                    if slot == microblock_slot && view_number == microblock.view_number() {
+                        let microheader2 = microblock.clone().unwrap_micro().header;
+                        let justification2 = block.clone().unwrap_micro().justification.signature;
+
+                        let proof = ForkProof {
+                            header1:        microheader1.clone(),
+                            header2:        microheader2.clone(),
+                            justification1: justification1,
+                            justification2: justification2};
```

Comment:
Avoid the cloning here.
```suggestion
                        let micro_block2 = microblock.unwrap_micro();
                        let microheader2 = micro_block2.header;
                        let justification2 = micro_block2.justification.signature;

                        let proof = ForkProof {
                            header1:        microheader1.clone(), // Previously, we only had a reference.
                            header2:        microheader2,
                            justification1: justification1.clone(), // Previously, we only had a reference.
                            justification2: justification2
                        };
```